XDMA/linux-kernel/xdma/xdma_netdev.c

```diff
@@ -0,0 +1,307 @@
+{
+	if (!engine) {
+		pr_err("Invalid DMA engine\n");
+		return -EINVAL; }
+	/* AXI ST or AXI MM non-incremental addressing mode? */
+	if (engine->non_incr_addr) {
+		int buf_lsb = (int)((uintptr_t)buf) & (engine->addr_align - 1);
+		size_t len_lsb = count & ((size_t)engine->len_granularity - 1);
+		int pos_lsb = (int)pos & (engine->addr_align - 1);
+
+		dbg_tfr("AXI ST or MM non-incremental\n");
+		dbg_tfr("buf_lsb = %d, pos_lsb = %d, len_lsb = %ld\n", buf_lsb,
+			pos_lsb, len_lsb);
+
+		if (buf_lsb != 0) {
+			dbg_tfr("FAIL: non-aligned buffer address %p\n", buf);
+			return -EINVAL;
+		}
+
+		if ((pos_lsb != 0) && (sync)) { dbg_tfr("FAIL: non-aligned AXI MM FPGA addr 0x%llx\n",
+				(unsigned long long)pos);
+			return -EINVAL;
+		}
+
+		if (len_lsb != 0) {
+			dbg_tfr("FAIL: len %d is not a multiple of %d\n",
+				(int)count,
+				(int)engine->len_granularity);
+			return -EINVAL;
+		}
+		/* AXI MM incremental addressing mode */
+	} else {
+		int buf_lsb = (int)((uintptr_t)buf) & (engine->addr_align - 1);
+		int pos_lsb = (int)pos & (engine->addr_align - 1);
+
+		if (buf_lsb != pos_lsb) {
+			dbg_tfr("FAIL: Misalignment error\n");
+			dbg_tfr("host addr %p, FPGA addr 0x%llx\n", buf, pos);
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+static void char_sgdma_unmap_kernel_buf(struct xdma_io_cb *cb, bool write)
+{
+	int i;
+
+	sg_free_table(&cb->sgt);
+        if (!cb->pages || !cb->pages_nr)
+		return;
+
+	for (i = 0; i < cb->pages_nr; i++) {
+		if (cb->pages[i]) {
+			if (!write)
+				set_page_dirty_lock(cb->pages[i]);
+			put_page(cb->pages[i]);
+		} else
+			break;
+	}
+
+	if (i != cb->pages_nr)
+		pr_info("sgl pages %d/%u.\n", i, cb->pages_nr);
+
+	kfree(cb->pages);
+	cb->pages = NULL;
+}
+
+static int char_sgdma_map_kernel_buf_to_sgl(struct xdma_io_cb *cb, bool write)
+{
+	struct sg_table *sgt = &cb->sgt;
+	unsigned long len = cb->len;
+	void __kernel *buf = cb->buf;
+	struct scatterlist *sg;
+	unsigned int pages_nr = (((unsigned long)buf + len + PAGE_SIZE - 1) -
+				 ((unsigned long)buf & PAGE_MASK))
+				>> PAGE_SHIFT;
+	int i;
+	int rv;
+        struct kvec kv;
+        memset(&kv, 0, sizeof(kv));
+        kv.iov_base = (void *)cb->buf;
+        kv.iov_len = PAGE_SIZE;
+	if (pages_nr == 0)
+		return -EINVAL;
+
+	if (sg_alloc_table(sgt, pages_nr, GFP_KERNEL)) {
+		pr_err("sgl OOM.\n");
+		return -ENOMEM;
+	}
+
+	cb->pages = kcalloc(pages_nr, sizeof(struct page *), GFP_KERNEL);
+	if (!cb->pages) {
+		pr_err("pages OOM.\n");
+		rv = -ENOMEM;
+		goto err_out;
+	}
+        rv = get_kernel_pages(&kv, pages_nr, 1, cb->pages);
+	/* No pages were pinned */
+	if (rv < 0) {
+		pr_err("unable to pin down %u kernel pages, %d.\n",
+			pages_nr, rv);
+		goto err_out;
+	}
+	/* Less pages pinned than wanted */
+	if (rv != pages_nr) {
+		pr_err("unable to pin down all %u kernel pages, %d.\n",
+			pages_nr, rv);
+		cb->pages_nr = rv;
+		rv = -EFAULT;
+		goto err_out;
+	}
+
+	for (i = 1; i < pages_nr; i++) {
```

Comment:
Please comment that why this need to be start from 1 instead of 0 and iterate over pages_nr - 1 times